File: computar_fibonacci/tutorial/C/fibonacci.c
#include <stdio.h>

int fib1(int n) {
  if(n == 1) {
    return 0;
  } else if(n == 2) {
    return 1;
  } else {
    return fib1(n-1) + fib1(n-2);
  }
}

int fib2(int n) {
  if(n == 1) {
    return 0;
  } else if(n == 2) {
    return 1;
  } else {
    // Arreglo para "recordar" los numeros de Fibonacci ya calculados.
    int memoria[n];
    memoria[0] = 0;
    memoria[1] = 1;
    int i;
    for(i = 2; i < n; i++) {
      memoria[i] = memoria[i-1] + memoria[i-2];
    }
    // Recordar que los arreglos de C comienzan desde el 0.
    return memoria[n-1];
  }
}

// Diálogo con el usuario.
int main(void) {
  printf("Ingrese un número entero: ");
  int number; // Declaramos la variable que vamos a pedir.
  scanf("%i", &number);
  printf("El n-ésino término de la serie de Fibonacci es:\n");
  printf("Calculado con fib1(n): ");
  printf("%i%s", fib1(number),"\n");
  printf("Calculado con fib2(n): ");
  printf("%i%s", fib2(number),"\n");
  // Los resultados son identicos, lo que cambia son los tiempos (comentar altenativamente entre implementaciones fib1 y fib2 para notarlo).
  return 0;
}

File: computar_fibonacci/tutorial/Ruby/fibonacci.rb
class Fibonacci

  def initialize()
    @fibonacci = 0
  end

  def fibonacci()
    puts @fibonacci
  end

  def fib1(n)
    if n == 1
      @fibonacci = 0
    elsif n == 2
      @fibonacci = 1
    else
      @fibonacci = fib1(n-1) + fib1(n-2)
    end
  end

  def fib2(n)
    if n == 1
      @fibonacci = 0
    elsif n == 2
      @fibonacci = 1
    else
      memory = Array.new()
      memory[0] = 0
      memory[1] = 1
      for i in 2..n-1
        memory[i] = memory[i-1] + memory[i-2]
      end
      @fibonacci = memory[n-1]
    end
  end

end

class Main
  # Diálogo con el usuario.
  def run()
    print "Ingrese un número entero: "
    n = gets.to_i

    print "El n-ésimo término de la serie de Fibonacci es: \n"

    fibonacci = Fibonacci.new()

    fibonacci.fib1(n)
    print "Calculado con fib1(n): "
    fibonacci.fibonacci()

    fibonacci.fib2(n)
    print "Calculado con fib2(n): "
    fibonacci.fibonacci()
  end
end

# Obs: La clase main y esta condicional, es simplemente para tener un estilo mas orientado a objetos.
if __FILE__ == $0
  x = Main.new()
  x.run()
end
